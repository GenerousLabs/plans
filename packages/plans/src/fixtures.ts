import mockFs from 'mock-fs';

export const bob = {
  slug: 'bob',
  data: { name: 'Bob' },
  frontmatter: `---
name: Bob
---`,
  markdown: '',
};
export const charlie = {
  slug: 'charlie',
  data: { name: 'Charlie' },
  frontmatter: `---
name: Charlie
---`,
  markdown: '',
};
export const aliceMessage = {
  data: { sender: 'alice', dateTimestampSeconds: 1590595620 },
  frontmatter: `---
sender: alice
dateTimestampSeconds: 1590595620
---`,
  markdown: 'Are you up for sharing this with me?',
};
export const charlieMessage = {
  data: { sender: 'charlie', dateTimestampSeconds: 1590595860 },
  frontmatter: `---
sender: charlie
dateTimestampSeconds: 1590595860
---`,
  markdown: "Would love to share this with you if you're interested.",
};

export const spotify = {
  slug: 'spotify',
  data: { name: 'Spotify' },
  frontmatter: `---
name: Spotify
---`,
  markdown: `I have two spots on my family plan.
`,
};
export const nordvpn = {
  slug: 'nordvpn',
  data: { name: 'NordVPN' },
  frontmatter: `---
name: NordVPN
---`,
  markdown: `I can use 6 devices at the same time
`,
};
export const omgyes = {
  slug: 'omgyes',
  data: { name: 'OMG Yes' },
  frontmatter: `---
name: OMG Yes
---`,
  markdown: `OMG Yes subscription
`,
};

export const joinFrontmatter = ({
  frontmatter,
  markdown,
}: {
  frontmatter: string;
  markdown: string;
}) => {
  return `${frontmatter}\n${markdown}`;
};

export const mockFilesystem = () => {
  mockFs({
    'alice/': {
      [bob.slug]: {
        'index.md': joinFrontmatter(bob),
        plans: {
          [spotify.slug]: {
            'index.md': joinFrontmatter(spotify),
          },
          [nordvpn.slug]: {
            'index.md': joinFrontmatter(nordvpn),
          },
        },
      },
      charlie: {
        'index.md': joinFrontmatter(charlie),
        plans: {
          [omgyes.slug]: {
            'index.md': joinFrontmatter(omgyes),
            [`message-${aliceMessage.data.dateTimestampSeconds}.md`]: joinFrontmatter(
              aliceMessage
            ),
            [`message-${charlieMessage.data.dateTimestampSeconds}.md`]: joinFrontmatter(
              charlieMessage
            ),
          },
        },
      },
    },
  });
};

export const mockFilesystemRestore = () => {
  mockFs.restore();
};
