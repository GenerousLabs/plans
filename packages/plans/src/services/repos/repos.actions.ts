import { GitParams } from '../../shared.types';
import { AppThunk } from '../../store';
import { noop, upsertOne } from './repos.state';
import { rootPathToMeRepoPath, updateRepo } from './repos.service';
import { to } from '../../utils/to.util';

export const init = ({
  fs,
  http,
  headers,
  rootPath,
}: GitParams & {
  rootPath: string;
}): AppThunk => async dispatch => {
  dispatch(
    noop({
      code: '#FmqCCM',
      message: 'init() start',
      params: { rootPath },
    })
  );

  const mePath = rootPathToMeRepoPath({ rootPath });

  const updateResponse = await to(
    updateRepo({ fs, http, headers, dir: mePath })
  );

  if (updateResponse.error) {
    const { error } = updateResponse;
    dispatch(
      noop({
        code: '#DL1pTF',
        message: 'Error updating me repo',
        params: {
          mePath,
          error: { code: (error as any).code, message: error.message },
        },
      })
    );
    return;
  }

  const { result } = updateResponse;

  dispatch(
    upsertOne({
      id: 'root',
      path: mePath,
      currentHeadCommitHash: result.commitOidAfter,
      lastFetchTimestampSeconds: Math.round(Date.now() / 1e3),
    })
  );

  /**
   * - Find the connections file
   * - Parse it
   * - Get the list of local repos
   *   - Push the data into redux
   * - Iterate over each repo
   *   - Pull it
   *   - Trigger the plans scan
   */
};
