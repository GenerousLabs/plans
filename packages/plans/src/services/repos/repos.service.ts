import git from 'isomorphic-git';
import yaml from 'js-yaml';
import { join } from 'path';
import { GitParams } from '../../shared.types';
import { doesFileExist } from '../../utils/fs.utils';

const PATH_TO_ME_REPO = 'me';
const CONNECTIONS_FILE_NAME = 'connections.yaml';
const CONNECTIONS_FOLDER_NAME = 'connections';

/**
 * A connection object that links the root user's me repository to another
 * repository that is shared with another person.
 */
export type Connection = {
  id: string;
  name: string;
  repoFolder: string;
  credentials?: {
    username?: string;
    password?: string;
    headers?: {
      [x: string]: string;
    };
  };
};

const author = {
  name: 'plans',
};

export const rootPathToMeRepoPath = ({ rootPath }: { rootPath: string }) => {
  return join(rootPath, PATH_TO_ME_REPO);
};

export const rootPathToUserRepoPath = ({
  rootPath,
  repoFolder,
}: {
  rootPath: string;
  repoFolder: string;
}) => {
  return join(rootPath, CONNECTIONS_FOLDER_NAME, repoFolder);
};

export const getCurrentCommit = async ({ fs, dir }: GitParams) => {
  const status = await git.log({ fs, dir, depth: 1 });

  if (status.length === 0) {
    throw new Error('No log found. #kzBW9C');
  }

  return status[0];
};

export const updateRepo = async ({ fs, http, headers, dir }: GitParams) => {
  const commitBefore = await getCurrentCommit({ fs, http, dir });

  try {
    await git.pull({
      fs,
      http,
      headers,
      dir,
      author,
    });
  } catch (error) {
    console.error('pull #snOHUF', error);
    throw error;
  }

  const commitAfter = await getCurrentCommit({ fs, http, dir });

  return {
    commitOidBefore: commitBefore.oid,
    commitOidAfter: commitAfter.oid,
  };
};

export const addConnectionFileNameToPath = ({ path }: { path: string }) => {
  return join(path, CONNECTIONS_FILE_NAME);
};

export const getConnectionsFromRepo = async ({
  fs,
  dir,
}: Pick<GitParams, 'fs' | 'dir'>) => {
  const path = addConnectionFileNameToPath({ path: dir });

  if (!doesFileExist({ fs, path })) {
    throw new Error('Connections file not found. #jQrd0p');
  }

  const contents = await fs.promises.readFile(path, { encoding: 'utf-8' });

  const data: Connection[] = yaml.safeLoad(contents);

  return data;
};
