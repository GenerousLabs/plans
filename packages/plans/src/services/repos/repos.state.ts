import {
  createAction,
  createEntityAdapter,
  createSlice,
} from '@reduxjs/toolkit';
import { REDUX_ROOT_KEY } from '../../constants';
import { Repo } from '../../shared.types';
import { RootState } from '../../store';

export const REDUCER_KEY = 'repos' as const;

const getState = (state: RootState) => state[REDUX_ROOT_KEY][REDUCER_KEY];

type RepoMeta = {
  path: string;
  lastFetchTimestampSeconds: number;
  currentHeadCommitHash: string;
};
type RepoState = Repo & RepoMeta;

const reposAdapter = createEntityAdapter<RepoState>();

const reposSlice = createSlice({
  name: 'PLANS/repos',
  initialState: reposAdapter.getInitialState(),
  reducers: {
    upsertOne: reposAdapter.upsertOne,
    updateOne: reposAdapter.updateOne,
    removeOne: reposAdapter.removeOne,
  },
});

export const { upsertOne, updateOne, removeOne } = reposSlice.actions;

export default reposSlice.reducer;

export const noop = createAction(
  'PLANS/repos/noop',
  ({
    code,
    message,
    params,
  }: {
    code: string;
    message?: string;
    params?: any;
  }) => {
    return {
      payload: {
        code,
        message,
        params,
      },
    };
  }
);

const reposSelectors = reposAdapter.getSelectors();

export const selectAll = (state: RootState) =>
  reposSelectors.selectAll(getState(state));
export const selectById = (state: RootState, id: string) =>
  reposSelectors.selectById(getState(state), id);
