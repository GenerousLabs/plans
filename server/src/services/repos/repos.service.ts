/* eslint-disable @typescript-eslint/camelcase */
import { join } from "path";
import { CONFIG_FILE_NAME, REPOS_ROOT, ME_REPO_NAME } from "../../constants";
import * as yaml from "js-yaml";
import fs from "fs";
import { splitRepo } from "../../util/repoNames";
import logger from "../../util/logger";

export const getRepoPath = ({ org, repo }: { org: string; repo: string }) => {
  return join(REPOS_ROOT, org, repo);
};

export const addConfigFileToRepoPath = (repoPath: string) =>
  join(repoPath, CONFIG_FILE_NAME);

type Config = {
  private_token: string;
  sharing_token: string;
};
const requiredStringProps = ["private_token", "sharing_token"] as const;
const isConfig = (input: unknown): input is Config => {
  for (const prop of requiredStringProps) {
    if (typeof (input as Config)[prop] !== "string") {
      return false;
    }
  }
  return true;
};

export const getConfigForRepoPath = async (repoPath: string) => {
  try {
    const { org } = splitRepo(repoPath);

    const diskPath = getRepoPath({ org, repo: ME_REPO_NAME });
    const configPath = addConfigFileToRepoPath(diskPath);
    const configYaml = await fs.promises.readFile(configPath, {
      encoding: "utf8",
    });
    const config = yaml.safeLoad(configYaml);
    if (isConfig(config)) {
      return config;
    }
    logger.debug("Failed to get config #ofaEoN", {
      config,
      configYaml,
      configPath,
      diskPath,
    });
    // NOTE: After logging here, we want to throw. But we're inside a try /
    // catch, so the throw is outside the try / catch below. Counter intuitive.
  } catch (error) {
    throw new Error("Auth fail. #kdmkwX");
  }

  // NOTE: This is thrown after the `logger.debug()` executes above.
  throw new Error("Auth fail. #aRSzZB");
};

export const checkReadPermissions = async ({
  repoPath,
  token,
}: {
  repoPath: string;
  token: string;
}) => {
  const { private_token, sharing_token } = await getConfigForRepoPath(repoPath);

  if (token === private_token || token === sharing_token) {
    return true;
  }
  throw new Error("Auth failure. #qcSHbN");
};

export const checkWritePermissions = async ({
  repoPath,
  token,
}: {
  repoPath: string;
  token: string;
}) => {
  const { private_token } = await getConfigForRepoPath(repoPath);

  if (token === private_token) {
    return true;
  }
  throw new Error("Auth failure. #XEWrYH");
};
