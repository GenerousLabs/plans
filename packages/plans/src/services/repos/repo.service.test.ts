import LightningFS from '@isomorphic-git/lightning-fs';
import 'fake-indexeddb/auto';
import { beforeEach, describe, expect, it } from 'jest-without-globals';
import { connectionsYaml } from '../../fixtures';
import { FS } from '../../shared.types';
import { getConnectionsFromRepo } from './repos.service';

describe('repo service', () => {
  describe('getConnectionsFromRepo()', () => {
    let fs: FS;
    beforeEach(async () => {
      fs = new LightningFS('__plans_tests', { wipe: true });
      await fs.promises.mkdir('/empty');
      await fs.promises.mkdir('/repo');
      await fs.promises.writeFile('/repo/connections.yaml', connectionsYaml);
    });

    it('throws when connections.yaml not found #Fv6TVV', async () => {
      await expect(
        getConnectionsFromRepo({
          fs,
          dir: '/empty',
        })
      ).rejects.toThrow();
    });

    it('successfully reads and parses connections.yaml #7vAlyp', async () => {
      expect(
        await getConnectionsFromRepo({ fs, dir: '/repo' })
      ).toMatchSnapshot();
    });
  });
});
