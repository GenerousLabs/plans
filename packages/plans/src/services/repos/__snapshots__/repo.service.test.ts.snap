// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`repo service getConn successfully reads and parses connections.yaml #7vAlyp 1`] = `
Array [
  Object {
    "id": "bob",
    "name": "Bob",
    "repoFolder": "bob",
  },
  Object {
    "id": "charlie",
    "name": "Charlie",
    "repoFolder": "charlie",
  },
]
`;
